[B+C] Added a way to make use of async chunk IO. Adds BUKKIT-3340

ShadowDog007 opened the pull request:
The Issue:
Currently there is no way for developers to make use of CraftBukkits asynchronous chunk loader. All plugin managed chunk loads are forced to run in the main thread. Depending on disk IO capabilities this can cause noticeable decrease in TPS.

Justification for this PR:
This PR aims to allow developers to utilize CraftBukkits chunk loader to prevent this decrease in TPS by providing a way for them to queue chunks for loading and be notified (Via  a callable object) upon the chunk  being loaded.

PR Breakdown:
Added a method World.loadChunk(int, int, Runnable)

Relevant PR: https://github.com/Bukkit/CraftBukkit/pull/1154

JIRA: https://bukkit.atlassian.net/browse/BUKKIT-3340

Commit:
Added a way to make use of async chunk IO. Adds BUKKIT-3340

This added method provides a way for developers to take advantage of
CraftBukkits asynchronous chunk loader

riking commented on src/main/java/org/bukkit/World.java:
```diff
@@ -173,6 +173,21 @@
     public void loadChunk(int x, int z);
 
     /**
+     * Loads the {@link Chunk} at the specified coordinates then calls the runnable object
```
Reword the end of this line

ShadowDog007 replied:
Eh. Why?

riking replied:
It sounds weird

ShadowDog007 replied:
My english is not that great. I can't think of another way to say it. :(

riking commented on src/main/java/org/bukkit/World.java:
```diff
@@ -173,6 +173,21 @@
+     * Loads the {@link Chunk} at the specified coordinates then calls the runnable object
+     * <p>
+     * If the chunk exists and is loaded the runnable object will be called straight away.
+     * If the chunk exists and is not loaded, it will be loaded asynchronously and the runnable object will be called in sync once it is done
```
78-column javadoc line wrapping, please.

ShadowDog007 replied:
I thought I was not supposed to? "No 80 column limit or 'weird' midstatement newlines."

natemort replied:
That's for code.  We have a different set of guidelines for Javadocs that are currently only on the wiki(http://wiki.bukkit.org/Getting_Involved).  They will be added to CONTRIBUTING.md sometime in the future.

ShadowDog007 replied:
Okies, I will change that then, if I can figure out how to change the commit......Q_Q

riking replied:
Just make a new one and upload to the `asyncChunkLoadingAPI` branch, we can handle the rebasing later.

ShadowDog007 replied:
Oh. Awesome. :)

riking commented on src/main/java/org/bukkit/World.java:
```diff
@@ -173,6 +173,21 @@
+     * If the chunk exists and is not loaded, it will be loaded asynchronously and the runnable object will be called in sync once it is done
+     * If the chunk does not exist it will be generated in the main thread
+     *
+     * If the given runnable is null the chunk is loaded synchronously
```
This seems like odd behavior and violates the POLS. I would think that plugins using an alternate method to load the chunk asynchronously but don't need a callback would want to pass in null.

ShadowDog007 replied:
Loading a chunk asynchronously without a callback does not provide much use. The chunk being loaded would likely be unloaded automatically before the plugin would require it.
I simply described the behavior of the method which I called within CraftBukkit

mbax replied:
Isn't there an event for when a chunk's loaded?

riking replied:
POLS = **P**rinciple **O**f **L**east **S**uprise
If we add a new method for async chunk loads, then having `null` as the last argument make the new functionality not happen is certainly suprising.

ShadowDog007 replied:
@mbax well, its a waste of timing listening for ChunkLoadEvents wating for a specific chunk to load when you can just use a runnable.

@riking Well, thats not due to my code, thats the way CraftBukkit devs implemented the chunk loader. So I thought I would just leave it.

Commit:
Added a way to make use of async chunk IO. Adds BUKKIT-3340

This added method provides a way for developers to take advantage of
CraftBukkits asynchronous chunk loader

ST-DDT commented:
It would be great, if you can load more than one chunk async and then call a runnable.
public void loadChunkWithCallback(Chunk[] chunks, Runnable runnable); 

And i think this method should be renamed (like the schedulledAsyncTask) to 
loadChunkAsynchroniosly(int ...
or 
loadChunkAsynchronioslyWith(Sync)Callback(int ...
This will tell the user what it does before reading the javadocs.

Btw can loadChunk be called from an async task?

ShadowDog007 commented:
@ST-DDT 
That name would be misleading as it does not always load the chunk asynchronously.
If the chunk needs to be generated it is done in the main thread.

To allow loading a set of chunks then calling a runnable would be too complicated.
Craftbukkit's already implemented system allows for a set of runnable objects per chunk.
That how I think it should be left.

If a plugin developer wanted to load a set of chunks, they should load them, and log the ones which have finished until they are all done while making sure they are not unloaded by Minecraft.

It has to be called from the main thread.

Commit:
Changed method to require a plugin reference.

Commit:
Changed method to use a 'QueuedProcess<Chunk>'

Commit:
Changed <Chunk> to <? super Chunk> + removed unneeded newline

riking commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * To be used when you wish an object to be queued for loading at a later time
+ */
+public interface QueuedProcess<T> {
```
How do I create a QueuedProcess? How do I satisfy the interface contract? Do I just subclass it like Runnable or Callable? :confused:

riking replied:
What am I supposed to do for isCancelled()? .... is it a way for me to say "No, I don't want to run after all"?

ShadowDog007 replied:
You create the object like you would any other object which implements something.

isCancceled() is just a way to make craftbukkit ignore the callback once its finished loading the chunk.

Check line 60 of this https://github.com/ShadowDog007/LonelySpawn/blob/master/src/com/forgenz/lonelyspawn/util/PlayerSpawnChecker.java

In my example it waits for a chunk to load before it 'spawns' a player. If the player logs off between clicking respawn and actually being teleported to the spawn then there is no point in having the callback run.

turt2live replied:
riking means "add this to the javadoc"

riking replied:
I just realized that I actually need this class, for another plugin. For a use unrelated to async chunk io - async database IO.

ShadowDog007 replied:
@turt2live @riking 
Added a bit more detail to the JavaDoc for 'isCancelled()'

Commit:
Added more detail to javadoc for QueuedProcess<T>

riking commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,22 @@
+
+
+/**
+ * To be used when you wish an object to be queued for loading at a later time
```
``` java
/**
 * A process to be called later which takes an argument. Typically, the
 * provided object will be created by an asynchronous procedure, and the
 * QueuedProcess will be ran when the object is ready for use.
 */
```

ShadowDog007 replied:
@riking
Added this, thanks :)

riking commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,22 @@
+     * want to receive the object
+     * @return True if the process is cancelled
+     */
+    public boolean isCancelled();
```
Honestly, I think this method is useless from a practical standpoint. Most all use-cases of it can be replaced by instant returning from the accept() method, and this shows in the CraftQueuedProcess wrapper implementation. If you want to 'cancel' it, you just instantly return from the accept() method.

ShadowDog007 replied:
@riking 
True.. I guess I just thought it would be a good idea to kinda, force developers to have a fail safe, so errors don't start popping up.

Removed it.

riking commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,22 @@
+public interface QueuedProcess<T> {
+    /**
+     * Called when the object requested has been fetched
+     * @param object The object which was queued to be fetched
```
Javadoc formatting: Please insert a line with just `*` on it before the `@param`.

Commit:
Give the annotations some space

Commit:
Removed isCancelled method

Commit:
Changed JavaDoc for QueuedProcess

turt2live commented on src/main/java/org/bukkit/World.java:
```diff
@@ -173,6 +174,28 @@
     public void loadChunk(int x, int z);
 
     /**
+     * Loads the {@link Chunk} at the specified coordinates then runs the
```
This entire JavaDoc looks like it needs to be reworded. Punctuation and more complete sentences should be used as this currently is more or less in bullet point form with periods.

There should be zero confusion to a plugin author as to what this does, what can and cannot happen, and what happens if something goes wrong (eg: a null callback). The author should also be able to see what is REQUIRED and what is not required (such as in the case of the callback - What happens if I give it null? Is it required?).

Is it also the best idea to notify the callback if the chunk is already loaded? I personally would rather see some sort of error or a clear indication that I, as an author, am trying to load something that already exists. (Or should we silently fail like in the existing load chunk method?)

turt2live commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,16 @@
+package org.bukkit.util;
+
+
```
Extra line... remove it please

turt2live commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,16 @@
+
+
+/**
+ * A process to be called later which takes an argument. Typically, the
```
This also needs rewording. (see https://github.com/Bukkit/Bukkit/pull/872/files#r6367219) 

Why would it "typically" be used for async procedures? There are a lot more questions that need to be answered by this JavaDoc.

ShadowDog007 replied:
I thought that was rather obvious. If you want something done asynchronously you usually have to wait for it. If it was something you could do in the main thread, you would just run the process it at the moment you want it.

turt2live commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,16 @@
+ * provided object will be created by an asynchronous procedure, and the
+ * QueuedProcess will be ran when the object is ready for use.
+ */
+public interface QueuedProcess<T> {
```
Do we need a second "failure" method? Something to be called if something goes wrong?

riking replied:
Man, wouldn't it be nice if [this](http://download.java.net/jdk8/docs/api/java/util/function/Consumer.html) was already in the default libraries? Oh well, go with what we got.

I already asked @ShadowDog007 to remove another second method, `isCancelled()`, on the grounds that it is easily replicated by a simple `if (condition) return;` at the beginning of accept(): https://github.com/Bukkit/Bukkit/pull/872#discussion_r5047367.

Also, exactly _what do you expect to go wrong_ in loading a chunk (or generating it, on the main thread, if it does not exist)? If we get an `IOException` during a chunk load, there's some bigger problem going on that a plugin really shouldn't be worrying about.

turt2live replied:
A plugin constantly waiting for `accept` to be called would cause some overhead. If the method is not going to be called, then there should, imo, be a method that can be called to tell the plugin that "this is never going to happen". Cases such as SOMETHING goes wrong in the generation/loading process (IO Error, whatever) would cause `accept` to never be called.

Edit: And although the plugin should not have to worry about it, they should at least get some notification that the requested chunk is not going to happen

turt2live commented on src/main/java/org/bukkit/util/QueuedProcess.java:
```diff
@@ -0,0 +1,16 @@
+     *
+     * @param object The object which was queued to be fetched
+     */
+    public void accept(T object);
```
The name of this doesn't seem right to me... suggestions?

riking replied:
The upcoming Java 8 standard library agrees with this wording: http://download.java.net/jdk8/docs/api/java/util/function/Consumer.html#accept(T)

turt2live replied:
We do not necessarily follow Java 8 standards. We generally follow the [Sun/Oracle Standards](http://www.oracle.com/technetwork/java/javase/documentation/codeconvtoc-136057.html) as per the CONTRIBUTING.md

Edit: Revised wording

Wolvereness replied:
0.o

Javadocs shouldn't be capitalized for the param, but Java 8 having this _exact_ method for this _exact_ purpose is a _pretty good_ justification.

turt2live replied:
well, in that case: leave it as is...

turt2live commented:
@ShadowDog007 If you are still interested in this please update it to the latest code base. The mentioned concerns will also need to be addressed. 

Once this is completed the review process may continue.

Commit:
Removed extra space in QueuedProcess.java

ShadowDog007 commented:
@turt2live 

How do I do that without having it adding every single commit in the last 4 months showing up in the commit? >.<
I've used this https://help.github.com/articles/syncing-a-fork with very little success.

Commit:
Merge remote-tracking branch 'upstream/master' into asyncChunkLoadingAPI

ShadowDog007 commented:
This is what I have, is this okay? http://gyazo.com/c39e4e77475ef071a47f7472164fb1be
Similar thing on both repo's

turt2live commented:
@ShadowDog007 So long as the "Commits" tab looks OK (no commits you didn't make) on here, then you're safe

ShadowDog007 commented:
Done :)

turt2live commented:
All it needs now is to address the mentioned concerns and to meet the CONTRIBUTING.md guidelines (namely the testing materials)

ShadowDog007 commented:
Already have this: https://github.com/Bukkit/CraftBukkit/pull/1154#issuecomment-18903678

turt2live commented:
It appears that the update process you did was not fully completed... 

The actual git log does not show the correct commits (meaning the applicable commits are not on top).

Please resolve this.

ShadowDog007 commented:
@turt2live 
Pretty sure it doesn't matter, don't you guys merge it all into one commit anyway? GitHub always shows commits in date order? I don't know how to change that...

turt2live commented:
We do, but we don't generally fix the issue of the log being screwed up.

What's happened with your PR is essentially this:

```
Your Commit
Bukkit commit
Bukkit commit
Bukkit commit
etc...
Your Commit
```

when it should be

```
Bukkit Commits
Your Commits
```

ShadowDog007 commented:
Well, I have no clue how to do that.

riking commented:
Try a `git rebase -i master` after you update `master` to the state of the main repo.

md-5 commented:
It looks completely usable to me @turt2live, once you pull it, just do a soft reset, and then rebase yourself. Conflicts will be minimal and I'm sure you can take the 2 minutes to do it yourself, since this fine gentlemen clearly can't.

turt2live commented:
Thanks md_5. I'll try it when I'm at a real computer again.

sent from mobile
On Sep 18, 2013 3:46 AM, "md-5" notifications@github.com wrote:

> It looks completely usable to me @turt2live https://github.com/turt2live,
> once you pull it, just do a soft reset, and then rebase yourself. Conflicts
> will be minimal and I'm sure you can take the 2 minutes to do it yourself,
> since this fine gentlemen clearly can't.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Bukkit/Bukkit/pull/872#issuecomment-24652604
> .

EdGruberman commented:
@ShadowDog007 You can clean up the log by rebasing your reference branch (master) locally (if you haven't already) then pushing that to your github repo.  [It looks like that](https://github.com/ShadowDog007/Bukkit/commits/master) is still out of date and why the log seems odd.

ShadowDog007 commented:
@EdGruberman No, the PR is on a different branch. https://github.com/ShadowDog007/Bukkit/commits/asyncChunkLoadingAPI

EdGruberman commented:
@ShadowDog007 Right, guessing from the comment on [that last commit on asyncChunkLoadingAPI](https://github.com/ShadowDog007/Bukkit/commit/17c8e146084004733da32e2e0fb3ca70f700c25b), it looks like you merged from upstream/master.  I assume your upstream remote is probably the same as your github master branch.  Also you should be rebasing instead of merging.

ShadowDog007 commented:
@EdGruberman Oh, I haven't had too much experience with GIT. I love Java. GIT makes me cry.

How exactly would I go about fixing that?

turt2live commented:
Okay, it looks fine on my end now :)

ghost commented:
_looks fine to turt2live and is ready to pull_
_epic 5 month freeze_

turt2live commented:
https://github.com/Bukkit/CraftBukkit/commit/b8fc6ab2c12e9b4c8d7b5370e44f23cc838014b2

Might have caused a problem...

@ShadowDog007 - Can you verify this PR is still needed and working? Thanks :D

ghost commented:
Yay! The humble turt2live is back to handle the pull request!

Commit:
Merge branch 'master' into asyncChunkLoadingAPI

ShadowDog007 commented:
I don't think that commit will cause any issues. Commit merge was pretty clean. Going to test it soon.

ghost commented:
soooooooo another month of inactivity

ShadowDog007 commented:
Yay. @turt2live would you mind looking at the plugin I provided? (On other PR)
